api/envoy/extensions/transport_sockets/internal_upstream/v3/internal_upstream.proto

```diff
@@ -0,0 +1,68 @@
+syntax = "proto3";
+
+package envoy.extensions.transport_sockets.internal_upstream.v3;
+
+import "envoy/config/core/v3/base.proto";
+import "envoy/type/metadata/v3/metadata.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.transport_sockets.internal_upstream.v3";
+option java_outer_classname = "InternalUpstreamProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/transport_sockets/internal_upstream/v3;internal_upstreamv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Internal Upstream]
+// [#extension: envoy.transport_sockets.internal_upstream]
+
+// Configuration for the internal upstream address. An internal address defines
+// a loopback user space socket residing in the same proxy instance. This
+// extension allows passing additional structured state across the user space
+// socket in addition to the regular byte stream. The purpose is to facilitate
+// communication between filters on the downstream and the upstream internal
+// connections.
+//
+// Because the passthrough state is transferred once per the upstream
+// connection before the bytes payload, every passthrough filter state object
+// is included in the hash key used to select an upstream connection if it
+// implements a hashing interface.
+//
+// .. note::
+//
+//  Using internal upstream transport affects load balancing decisions if the
+//  passthrough state is derived from the downstream connection attributes. As
+//  an example, using the downstream source IP in the passthrough state will
+//  prevent sharing of an upstream internal connection between different source
+//  IPs.
+message InternalUpstreamTransport {
+  // Describes the location of the imported metadata value.
+  message MetadataValueSource {
+    // Specifies what kind of metadata.
+    type.metadata.v3.MetadataKind kind = 1;
+
+    // Name is the filter namespace used in the dynamic metadata.
+    string name = 2 [(validate.rules).string = {min_len: 1}];
+  }
+
+  // Describes the location of the imported filter state object from the downstream connection.
+  message FilterStateSource {
+    // Name is the imported filter state object name.
+    string name = 2 [(validate.rules).string = {min_len: 1}];
+  }
+
+  // Specifies the metadata namespaces and values to insert into the downstream
+  // internal connection dynamic metadata when an internal address is used as a host.
+  repeated MetadataValueSource passthrough_metadata = 1;
```

Comment:
Yes, it will override provided there is source. But it's possible to not have a source metadata. For example:
```
cluster1: Metadata[X=Y]
- endpoint1: Metadata[X=Z]
- endpoint2: -
```

If we specify as source:
- {kind: {cluster: {}}, name: X}
- {kind: {endpoint: {}}, name: X}

Then for endpoint1 the passthrough metadata is Z while for endpoint2 the passthrough metadata is Y.

I think your suggestion makes sense at runtime, we can resolve overrides at extraction.